lib/Dialect/FIRRTL/Transforms/SpecializeLayers.cpp

```diff
@@ -0,0 +1,725 @@
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+
+// TODO: this should be upstreamed.
+namespace llvm {
+template <>
+struct PointerLikeTypeTraits<mlir::ArrayAttr>
+    : public PointerLikeTypeTraits<mlir::Attribute> {
+  static inline mlir::ArrayAttr getFromVoidPointer(void *ptr) {
+    return mlir::ArrayAttr::getFromOpaquePointer(ptr);
+  }
+};
+} // namespace llvm
+
+namespace {
+/// Removes non-local annotations whose path is no longer viable, due to
+/// the removal of module instances.
+struct AnnotationCleaner {
+  /// Create an AnnotationCleaner which removes any annotation which contains a
+  /// reference to a symbol in removedPaths.
+  AnnotationCleaner(const DenseSet<StringAttr> &removedPaths)
+      : removedPaths(removedPaths) {}
+
+  AnnotationSet cleanAnnotations(AnnotationSet annos) {
+    annos.removeAnnotations([&](Annotation anno) {
+      return removedPaths.contains(
+          anno.getMember<FlatSymbolRefAttr>("circt.nonlocal").getAttr());
+    });
+    return annos;
+  }
+
+  void cleanAnnotations(Operation *op) {
+    AnnotationSet oldAnnotations(op);
+    // We want to avoid attaching an empty annotation array on to an op that
+    // never had an annotation array in the first place.
+    if (!oldAnnotations.empty())
+      cleanAnnotations(oldAnnotations).applyToOperation(op);
+  }
+
+  void operator()(FModuleLike module) {
+    // Clean the regular annotations.
+    cleanAnnotations(module);
+
+    // Clean all port annotations.
+    for (size_t i = 0, e = module.getNumPorts(); i < e; ++i) {
+      auto oldAnnotations = AnnotationSet::forPort(module, i);
+      auto newAnnotations = cleanAnnotations(oldAnnotations);
+      newAnnotations.applyToPort(module, i);
+    }
+
+    // Clean all annotations in body.
+    module->walk([&](Operation *op) {
+      // Clean regular annotations.
+      cleanAnnotations(op);
+
+      if (auto mem = dyn_cast<MemOp>(op)) {
+        // Update all annotations on ports.
+        for (size_t i = 0, e = mem.getNumResults(); i < e; ++i) {
+          auto oldAnnotations = AnnotationSet::forPort(mem, i);
+          auto newAnnotations = cleanAnnotations(oldAnnotations);
+          newAnnotations.applyToPort(mem, i);
+        }
+      }
+    });
+  }
+
+  /// A set of symbols of removed paths.
+  const DenseSet<StringAttr> &removedPaths;
+};
+
+/// Helper to keep track of an insertion point and move operations around.
+struct InsertionPoint {
+  /// Create an insertion point at the end of a block.
+  static InsertionPoint atBlockEnd(Block *block) {
+    return InsertionPoint{block, block->end()};
+  }
+
+  /// Move the target operation before the current insertion point and update
+  /// the insertion point to point to the op.
+  void moveOpBefore(Operation *op) {
+    op->moveBefore(block, it);
+    it = Block::iterator(op);
+  }
+
+private:
+  InsertionPoint(Block *block, Block::iterator it) : block(block), it(it) {}
+
+  Block *block;
+  Block::iterator it;
+};
```

Comment:
Everyone find this confusing but I have a suggestion from @rwy7 which will hopefully make it clearer, but I am starting to think it is not really an abstraction which makes sense to other people and I should delete it.  I was sort of thinking of this as an insertion point that points between operations instead of an operation.  Basically this is an insertion point that modifies itself every time you move an operation after it, to point to the moved operation.  The reason I need it is a little involved but it at a high level its because I am both enabling and disabling layers at the same time. When we are disabling layers, we have to delete operations, and so we walk the blocks in reverse to make sure that we removed uses of values before the value itself.  When we enable a layer, we inline the contents of the layerblock into the outer context, which turns out to be a little difficult to do in reverse.  When we inline an operation, we need to make sure that the insertion point is updated so that the next operation inlined goes before the previously inlined operation.